Moved duplicate author declarations to setUpTestData() in UpdateViewT…

MyungSeKyo opened the pull request:
…ests

MyungSeKyo commented:
@carltongibson 
I agree with you I changed it that way.
And I dont't know Why some test failure generated without `pk=1` in `Author.objects.create()`

carltongibson commented:
> And I don't know Why some tests fail without pk=1 in Author.objects.create()

I'm not seeing that. I remove the `pk=1` and all passes: 

<details>
<summary>Git diff & test run</summary>

```
(django) ~/Documents/Django-Stack/django/tests (pr/10618)$ git diff
diff --git a/tests/generic_views/test_edit.py b/tests/generic_views/test_edit.py
index cf80741fa5..5a0c998284 100644
--- a/tests/generic_views/test_edit.py
+++ b/tests/generic_views/test_edit.py
@@ -217,7 +217,6 @@ class UpdateViewTests(TestCase):
     @classmethod
     def setUpTestData(cls):
         cls.author = Author.objects.create(
-            pk=1,
             name='Randall Munroe',
             slug='randall-munroe',
         )
(django) ~/Documents/Django-Stack/django/tests (pr/10618)$ ./runtests.py generic_views.test_edit.UpdateViewTests
Testing against Django installed in '/Users/carlton/Documents/Django-Stack/django/django' with up to 4 processes
Creating test database for alias 'default'…
Creating test database for alias 'other'…
System check identified no issues (0 silenced).
........
----------------------------------------------------------------------
Ran 8 tests in 0.201s

OK
Destroying test database for alias 'default'…
Destroying test database for alias 'other'…
```

</details>

<br>
Can you show me your failure?

MyungSeKyo commented:
Before I fixed my commit, There was no `pk=1` and the `UpdateViewTests.test_update_get_object` was failed in `pull-requests-bionic/database=postgres,label=bionic-pr,python=pythonX.X`.
because `{DoesNotExist}Author matching query does not exist.`
So I added `pk=1` and there are no failures

carltongibson commented:
Looks like it's from https://code.djangoproject.com/ticket/15544 https://github.com/django/django/commit/806bffcf08893fc9a5104b3a1208551a5627473e. 

And the view code is using a hard-coded `pk`, so that makes sense: 

```
class OneAuthorUpdate(generic.UpdateView):
    success_url = '/list/authors/'
    fields = '__all__'

    def get_object(self):
        return Author.objects.get(pk=1)
```

(generic_views/views.py:160)

MyungSeKyo commented:
Oh I see.
I removed the commit for showing failure
and remain the `pk=1` parameter.

felixxm commented:
Sorry, closed by mistake.

timgraham commented on tests/generic_views/test_edit.py:
```diff
@@ -214,34 +214,37 @@ class MyCreateView(CreateView):
+    @classmethod
+    def setUpTestData(cls):
+        cls.author = Author.objects.create(
+            pk=1,
```
Since hardcoded pks should be avoided, add a comment:
```
pk=1,  # Required for OneAuthorUpdate.
```

timgraham commented on tests/generic_views/test_edit.py:
```diff
@@ -214,34 +214,37 @@ class MyCreateView(CreateView):
         self.assertIsInstance(res.context['form'], forms.ModelForm)
-        self.assertEqual(res.context['object'], Author.objects.get(pk=a.pk))
-        self.assertEqual(res.context['author'], Author.objects.get(pk=a.pk))
+        self.assertEqual(res.context['object'], Author.objects.get(pk=self.author.pk))
```
I don't see the need to refetch the object from the database.
`self.assertEqual(res.context['object'], self.author)`
should work fine for all these assertions. Maybe the original test author didn't realize that model equality only compares primary keys.

timgraham commented on tests/generic_views/test_edit.py:
```diff
@@ -214,34 +214,37 @@ class MyCreateView(CreateView):
-            slug='randall-munroe',
-        )
-        res = self.client.post('/edit/author/%d/update/' % a.pk, {'name': 'A' * 101, 'slug': 'randall-munroe'})
+        res = self.client.post('/edit/author/%d/update/' % self.author.pk, {
```
I guess you should follow the indentation of the other tests.

MyungSeKyo commented:
@timgraham Thanks for your review
I fixed code according to your review and squashed into one commit.
